executors/cmdexec/volume.go

```diff
@@ -78,6 +79,99 @@ func (s *CmdExecutor) VolumeCreate(host string,
 	return &executors.Volume{}, nil
 }
 
+func (s *CmdExecutor) isRemoveBrickComplete(commands []string, host string) bool {
+	type RemoveBrickStatusOutput struct {
+		OpRet          int    `xml:"opRet"`
+		OpErrno        int    `xml:"opErrno"`
+		OpErrStr       string `xml:"opErrstr"`
+		VolRemoveBrick struct {
+			Aggregate struct {
+				StatusStr string `xml:"statusStr"`
+			} `xml:"aggregate"`
+		} `xml:"volRemoveBrick"`
+	}
+	statusOutputXml, err := s.RemoteExecutor.RemoteCommandExecute(host, commands, 10)
+	if err != nil {
+		logger.LogError("Unable to check remove-brick status")
+		return false
+	}
+	var statusOutput RemoveBrickStatusOutput
+	err = xml.Unmarshal([]byte(statusOutputXml[0]), &statusOutput)
+	if err != nil {
+		logger.LogError("Unable to determine remove-brick status output")
+		return false
+	}
+	logger.Debug("%+v\n", statusOutput)
+	if statusOutput.VolRemoveBrick.Aggregate.StatusStr != "completed" {
+		return false
+	}
+
+	return true
+}
+
+func (s *CmdExecutor) cleanupAddBrick(volume *executors.VolumeRequest, host string, inSet, maxPerSet int) error {
+	// If add-brick or rebalance has failed, we have to remove the bricks using the following process
+	// 1. gluster volume remove-brick VOLNAME BRICKs start
+	// 2. gluster volume remove-brick VOLNAME BRICKs status and grep for all completed
+	// 3. gluster volume remove-brick VOLNAME BRICKs commit
+
+	type RemoveBrickStartCommitOutput struct {
+		OpRet    int    `xml:"opRet"`
+		OpErrno  int    `xml:"opErrno"`
+		OpErrStr string `xml:"opErrstr"`
+	}
+	var removeBrickStatusComplete bool
+	commands := s.createRemoveBrickCommands(volume, 0, inSet, maxPerSet, "start --xml")
+	startOutputXML, err := s.RemoteExecutor.RemoteCommandExecute(host, commands, 10)
+	if err != nil {
+		return fmt.Errorf("Could not revert add-brick")
+	}
+	var startOutput RemoveBrickStartCommitOutput
+	err = xml.Unmarshal([]byte(startOutputXML[0]), &startOutput)
+	if err != nil {
+		logger.LogError("Unable to determine remove-brick start output")
+		return fmt.Errorf("Could not revert add-brick")
+	}
+	logger.Debug("%+v\n", startOutput)
+	if startOutput.OpRet != 0 {
+		logger.LogError("failed to start remove-brick, manual cleanup might be required")
+		return fmt.Errorf("Could not revert add-brick")
+	}
+
+	commands = s.createRemoveBrickCommands(volume, 0, inSet, maxPerSet, "status --xml")
+	ticker := time.NewTicker(time.Second * 10)
+	select {
+	case <-ticker.C:
+		removeBrickStatusComplete = s.isRemoveBrickComplete(commands, host)
+		if removeBrickStatusComplete {
+			ticker.Stop()
+			commands := s.createRemoveBrickCommands(volume, 0, inSet, maxPerSet, "commit --xml")
+			commitOutputXML, err := s.RemoteExecutor.RemoteCommandExecute(host, commands, 10)
```

Comment:
It seems to me that all the command execution and xml parsing should be rolled into yet another helper func.